src/codeeditor/editor.ts

```diff
@@ -0,0 +1,804 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import {
+  IDisposable
+} from 'phosphor/lib/core/disposable';
+
+import {
+  Message
+} from 'phosphor/lib/core/messaging';
+
+import {
+  ISignal, clearSignalData, defineSignal
+} from 'phosphor/lib/core/signaling';
+
+import {
+  Widget
+} from 'phosphor/lib/ui/widget';
+
+import {
+  IChangedArgs
+} from '../common/interfaces';
+
+
+/**
+ * A namespace for code editors.
+ */
+export
+namespace CodeEditor {
+  /**
+   * A zero-based position in the editor.
+   */
+  export
+  interface ICursorPosition {
+    /**
+     * The cursor line number.
+     */
+    line: number;
+
+    /**
+     * The cursor column number.
+     */
+    column: number;
+  }
+
+  /**
+   * An interface describing editor state coordinates.
+   */
+  export
+  interface ICoords {
+    /**
+     * The left coordinate value.
+     */
+    readonly left: number;
+
+    /**
+     * The right coordinate value.
+     */
+    readonly right: number;
+
+    /**
+     * The top coordinate value.
+     */
+    readonly top: number;
+
+    /**
+     * The bottom coordinate value.
+     */
+    readonly bottom: number;
+  }
+
+  /**
+   * Configuration options for the editor.
+   */
+  export
+  interface IConfiguration extends IDisposable {
+    /**
+     * A signal emitted when the configuration changes.
+     */
+    changed: ISignal<IConfiguration, string>;
+
+    /**
+     * Whether line numbers should be displayed. Defaults to false.
+     */
+    lineNumbers: boolean;
+
+    /**
+     * Set to false for horizontal scrolling. Defaults to true.
+     */
+    wordWrap: boolean;
+
+    /**
+     * Whether the editor is read-only.  Defaults to false.
+     */
+    readOnly: boolean;
+
+    /**
+     * The number of spaces a tab is considered equal to. Defaults to 4.
+     */
+    tabSize: number;
+
+    /**
+     * Translate tabs to spaces. Defaults to true.
+     */
+    tabsToSpaces: boolean;
+
+    /**
+     * Set to false to stop auto pairing quotes, brackets etc.
+     * Defaults to true.
+     */
+    autoMatchEnabled: boolean;
```

Comment:
I think it is important for the user code to have explicit undefined value, so it can decide about capabilities of an underlying editor, for example:
```typescript
if (editor.autoMatchEnabled) {
  // add a command and a menu item for auto match functionality
}
```